pbs/usersync.go

```diff
@@ -10,33 +10,86 @@ import (
 	"strings"
 	"time"
 
+	"errors"
 	"github.com/golang/glog"
 	"github.com/julienschmidt/httprouter"
 	"github.com/prebid/prebid-server/ssl"
-	metrics "github.com/rcrowley/go-metrics"
+	"github.com/rcrowley/go-metrics"
 )
 
-var cookie_domain string
-var external_url string
-var recaptcha_secret string
+// Recaptcha code from https://github.com/haisum/recaptcha/blob/master/recaptcha.go
+const RECAPTCHA_URL = "https://www.google.com/recaptcha/api/siteverify"
+const COOKIE_NAME = "uids"
 
-type PBSCookie struct {
-	UIDs     map[string]string `json:"uids,omitempty"`
-	OptOut   bool              `json:"optout,omitempty"`
-	Birthday *time.Time        `json:"bday,omitempty"`
+const (
+	USERSYNC_OPT_OUT     = "usersync.opt_outs"
+	USERSYNC_BAD_REQUEST = "usersync.bad_requests"
+	USERSYNC_SUCCESS     = "usersync.%s.sets"
+)
+
+type UserSyncDeps struct {
+	Cookie_domain    string
+	External_url     string
+	Recaptcha_secret string
+	Metrics          metrics.Registry
+}
+
+// UserSyncMap is cookie which stores the user sync info for all of our bidders.
+//
+// To get an instance of this from a request, use ParseUserSyncMapFromRequest.
+// To write an instance onto a response, use SetCookieOnResponse.
+type UserSyncMap interface {
```

Comment:
Good/fair question.

I really wanted to be _sure_, without having to look outside this file, that all the other code is going through the setter methods. The main value being that if someone else has to write cookie sync code somewhere in PBS, they won't have to know about this particular exception to "get it right."

So... I tried to make `UIDs` private, and add a the setter method to the `PBSCookie`. But it turns out that Golang's JSON marshal & unmarshal don't work on private variables.

This was the next best thing I could come up with. UIDs can stay public, but`cookieImpl` will be private. So the rest of PBS is forced to go through the setters.